File: ruby/merge_sort.rb
def merge_sort(arr)
  return arr if arr.length <= 1

  middle = arr.length / 2
  left = arr[0...middle]
  right = arr[middle..-1]

  merge(merge_sort(left), merge_sort(right))
end

def merge(left, right)
  result = []
  left_index = 0
  right_index = 0

  while left_index < left.length && right_index < right.length
    if left[left_index] < right[right_index]
      result << left[left_index]
      left_index += 1
    else
      result << right[right_index]
      right_index += 1
    end
  end

  result.concat(left[left_index..-1])
  result.concat(right[right_index..-1])

  result
end

# Example usage
arr = [64, 34, 25, 12, 22, 11, 90]
sorted_arr = merge_sort(arr)
puts "Sorted array: #{sorted_arr}"

File: php/merge_sort.php
<?php

function mergeSort($arr) {
    if (count($arr) <= 1) return $arr;

    $middle = floor(count($arr) / 2);
    $left = array_slice($arr, 0, $middle);
    $right = array_slice($arr, $middle);

    return merge(mergeSort($left), mergeSort($right));
}

function merge($left, $right) {
    $result = [];
    $leftIndex = $rightIndex = 0;

    while ($leftIndex < count($left) && $rightIndex < count($right)) {
        if ($left[$leftIndex] < $right[$rightIndex]) {
            $result[] = $left[$leftIndex++];
        } else {
            $result[] = $right[$rightIndex++];
        }
    }

    return array_merge($result, array_slice($left, $leftIndex), array_slice($right, $rightIndex));
}

// Example usage
$arr = [64, 34, 25, 12, 22, 11, 90];
$sortedArr = mergeSort($arr);
echo "Sorted array: ";
print_r($sortedArr);

?>
